### PR TITLE
Move to V4 API

### LINKED ISSUE #3: Update `go-gitlab` dependency to support /v4 APIs
_This issue was originally opened by @richardc as hashicorp/terraform#14823. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

The `gitlab` provider is based on the excellent [go-gitlab](https://github.com/xanzy/go-gitlab/) library.  At time of writing this is coded in terms of the [/v3 gitlab api](https://gitlab.com/gitlab-org/gitlab-ce/blob/8-16-stable/doc/api/README.md) though work is underway in the go-gitlab project to modernise that: https://github.com/xanzy/go-gitlab/issues/151  

This is significant as the v4 api has been considered stable and preferred since the release of gitlab 9.0 in March, and the v3 api is scheduled to be removed in gitlab 9.5 https://docs.gitlab.com/ce/api/v3_to_v4.html  It's also the case that the v4 api adds additional features which will be good to make use of.

Once the go-gitlab project releases api v4 support, bump the dependency, and recode impacted resources to embrace the new capabilities exposed.

### PR DESCRIPTION
**Pull request by roidelapluie**
Fixes #3

### CODE REVIEW

**Comment by roidelapluie** [comment]
```$ make testacc 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?       github.com/terraform-providers/terraform-provider-gitlab        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGitlabDeployKey_basic
--- PASS: TestAccGitlabDeployKey_basic (2.19s)
=== RUN   TestAccGitlabGroup_basic
--- PASS: TestAccGitlabGroup_basic (0.72s)
=== RUN   TestAccGitlabProjectHook_basic
--- PASS: TestAccGitlabProjectHook_basic (2.03s)
=== RUN   TestAccGitlabProject_basic
--- PASS: TestAccGitlabProject_basic (2.06s)
=== RUN   TestGitlab_validation
--- PASS: TestGitlab_validation (0.00s)
=== RUN   TestGitlab_visbilityHelpers
--- PASS: TestGitlab_visbilityHelpers (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-gitlab/gitlab 7.022s
```

**Comment by tombuildsstuff** [comment]
I think this needs to be merged after #19?

**Comment by roidelapluie** [comment]
#19 is included in this PR

#7 is not needed coz it has been fixed upstream in gitlab

thanks

**Comment by roidelapluie** [comment]
```
$ make testacc 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?       github.com/terraform-providers/terraform-provider-gitlab        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGitlabDeployKey_basic
--- PASS: TestAccGitlabDeployKey_basic (8.61s)
=== RUN   TestAccGitlabGroup_basic
--- PASS: TestAccGitlabGroup_basic (1.10s)
=== RUN   TestAccGitlabProjectHook_basic
--- PASS: TestAccGitlabProjectHook_basic (6.13s)
=== RUN   TestAccGitlabProject_basic
--- PASS: TestAccGitlabProject_basic (5.93s)
=== RUN   TestGitlab_validation
--- PASS: TestGitlab_validation (0.00s)
=== RUN   TestGitlab_visbilityHelpers
--- PASS: TestGitlab_visbilityHelpers (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-gitlab/gitlab 21.768s
```

**Comment by roidelapluie** [comment]
@tombuildsstuff rewrote this

**Comment by roidelapluie** [comment]
@tombuildsstuff rebased

```
$ make testacc 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?       github.com/terraform-providers/terraform-provider-gitlab        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGitlabDeployKey_basic
--- PASS: TestAccGitlabDeployKey_basic (6.69s)
=== RUN   TestAccGitlabGroup_basic
--- PASS: TestAccGitlabGroup_basic (0.79s)
=== RUN   TestAccGitlabProjectHook_basic
--- PASS: TestAccGitlabProjectHook_basic (5.95s)
=== RUN   TestAccGitlabProject_basic
--- PASS: TestAccGitlabProject_basic (5.91s)
=== RUN   TestGitlab_validation
--- PASS: TestGitlab_validation (0.00s)
=== RUN   TestGitlab_visbilityHelpers
--- PASS: TestGitlab_visbilityHelpers (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-gitlab/gitlab 19.333s
```

File: gitlab/resource_gitlab_project.go [hunk 1]
```diff
@@ -196,5 +195,35 @@ func resourceGitlabProjectDelete(d *schema.ResourceData, meta interface{}) error
 	log.Printf("[DEBUG] Delete gitlab project %s", d.Id())
 
 	_, err := client.Projects.DeleteProject(d.Id())
-	return err
+	if err != nil {
+		return err
+	}
+
+	// Wait for the datacenter resource to be ready
```
Comment (richardc) [inline gitlab/resource_gitlab_project.go]: Datacenter?  Is this bad copy and paste?

File: gitlab/resource_gitlab_project_hook.go [hunk 2]
```diff
@@ -89,7 +84,6 @@ func resourceGitlabProjectHookCreate(d *schema.ResourceData, meta interface{}) e
 		MergeRequestsEvents:   gitlab.Bool(d.Get("merge_requests_events").(bool)),
 		TagPushEvents:         gitlab.Bool(d.Get("tag_push_events").(bool)),
 		NoteEvents:            gitlab.Bool(d.Get("note_events").(bool)),
-		BuildEvents:           gitlab.Bool(d.Get("build_events").(bool)),
```
Comment (richardc) [inline gitlab/resource_gitlab_project_hook.go]: These got renamed to JobEvents which go-gitlab does support; https://github.com/xanzy/go-gitlab/blob/master/projects.go#L590

**Comment by roidelapluie** [comment]
```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 120m
?       github.com/terraform-providers/terraform-provider-gitlab        [no test files]
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGitlabDeployKey_basic
--- PASS: TestAccGitlabDeployKey_basic (6.05s)
=== RUN   TestAccGitlabGroup_basic
--- PASS: TestAccGitlabGroup_basic (0.59s)
=== RUN   TestAccGitlabProjectHook_basic
--- PASS: TestAccGitlabProjectHook_basic (5.83s)
=== RUN   TestAccGitlabProject_basic
--- PASS: TestAccGitlabProject_basic (5.94s)
=== RUN   TestGitlab_validation
--- PASS: TestGitlab_validation (0.00s)
=== RUN   TestGitlab_visbilityHelpers
--- PASS: TestGitlab_visbilityHelpers (0.00s)
PASS
ok      github.com/terraform-providers/terraform-provider-gitlab/gitlab 18.412s
```

**Pull request merged**